internal/util/streamrpc/streamer.go

```diff
@@ -133,11 +174,18 @@ func (s *ResultCacheServer) Send(result *internalpb.RetrieveResults) error {
 	}
 
 	s.cache.Put(result)
-	if s.cache.IsFull() {
+	if s.cache.IsFull() && s.cache.size <= s.maxMsgSize {
 		result := s.cache.Flush()
 		if err := s.srv.Send(result); err != nil {
 			return err
 		}
+	} else if s.cache.IsFull() && s.cache.size > s.maxMsgSize {
```

Comment:
Is it possible that we return the results capped with `maxMsgSize` in `s.cache.Flush()`, instead of splitting messages and send them separately?

We will introduce 2 problems by splitting messages: 1) the additional memory copy will harm the total performance, and 2) we'll send a small result, which is the last element of the return of the split, every time when cache is full.

Reply:
The split don't have large memory copy, it reference part of buffer result ids directly.